lua/neorg/modules/core/integrations/treesitter/module.lua

```diff
@@ -231,31 +231,51 @@ module.public = {
 
         -- Do we need to go through each tree? lol
         vim.treesitter.get_parser(opts.buf, opts.ft):for_each_tree(function(tree)
-            -- Get the root for that tree
-            ---@type TSNode
-            local root = tree:root()
+            table.insert(result, module.public.search_tree(tree, node_type))
+        end)
 
-            --- Recursively searches for a node of a given type
-            ---@param node TSNode #The starting point for the search
-            local function descend(node)
-                -- Iterate over all children of the node and try to match their type
-                for child, _ in node:iter_children() do ---@diagnostic disable-line -- TODO: type error workaround <pysan3>
-                    if child:type() == type then
-                        table.insert(result, child)
-                    else
-                        -- If no match is found try descending further down the syntax tree
-                        for _, child_node in ipairs(descend(child) or {}) do
-                            table.insert(result, child_node)
-                        end
+        return vim.iter(result):flatten():totable()
+    end,
+
+    ---Gets all nodes of a given type from the AST
+    ---@param node_type string #The type of node to filter out
+    ---@param path string path to the file to parse
+    ---@param filetype string? file type of the file or `norg` if omitted
+    get_all_nodes_in_file = function(node_type, path, filetype)
+        path = vim.fs.normalize(path)
+        if not filetype then filetype = "norg" end
+
+        local contents = io.open(path, "r"):read("*a")
+        local tree = vim.treesitter.get_string_parser(contents, filetype):parse()[1]
+        if not (tree or tree.root) then return {} end
+
+        return module.public.search_tree(tree, node_type)
+    end,
+
+    search_tree = function(tree, node_type)
+        local result = {}
+        local root = tree:root()
+
+        --- Recursively searches for a node of a given type
+        ---@param node userdata #The starting point for the search
+        local function descend(node)
+            -- Iterate over all children of the node and try to match their type
+            for child, _ in node:iter_children() do ---@diagnostic disable-line -- TODO: type error workaround <pysan3>
```

Comment:
suggestion: while we're at it, give this boy an upgrade, then you can remove the type error workaround a few line lower:
```suggestion
        ---@param node TSNode #The starting point for the search
        local function descend(node)
            -- Iterate over all children of the node and try to match their type
            for child, _ in node:iter_children() do
```